cpp/src/gandiva/parse_url_holder.h

```diff
@@ -111,45 +140,61 @@ namespace gandiva {
         ExecutionContext *ctx, const char * url, int32_t url_len,
         const char * part, int32_t part_len,
         const char * pattern, int32_t pattern_len, int32_t *out_length) {
-      std::string url_string(url, url_len);
       std::string part_string(part, part_len);
       std::string pattern_string(pattern, pattern_len);
       arrow::internal::Uri uri;
       std::string out;
 
-      auto status = uri.Parse(url_string);
-      if (!status.ok()) {
+      if (part_string != "QUERY") {
         return nullptr;
       }
 
-      if (part_string != "QUERY" || !uri.has_query()) {
-        return nullptr;
-      } else {
-        RE2 re2("(&|^)" + pattern_string + "=([^&]*)");
-        int groups_num = re2.NumberOfCapturingGroups();
-        RE2::Arg *args[groups_num];
-        for (int i = 0; i < groups_num; i++) {
-          args[i] = new RE2::Arg;
-        }
-        *(args[1]) = &out;
-        // Use re2 instead of pattern_ for better performance.
-        bool matched = RE2::PartialMatchN(uri.query_string(), re2, args, groups_num);
-        if (!matched) {
-          *out_length = 0;
+      // Here we skip the query parsing, as urlparser does not support invalid characters in url,
+      // which are actually supported in vanilla Spark, except spaces.
+      int query_start_idx = -1;
+      for (int idx = url_len - 1; idx >= 0; idx--) {
+        if (url[idx] == ' ') {
           return nullptr;
+        } else if (url[idx] == '?') {
+          query_start_idx = idx;
+          break;
         }
+      }
+      if (query_start_idx == -1) {
+        return nullptr;
+      }
 
-        *out_length = static_cast<int32_t>(out.length());
-        char *result_buffer = reinterpret_cast<char *>(ctx->arena()->Allocate(*out_length));
-        if (result_buffer == NULLPTR) {
-          ctx->set_error_msg("Could not allocate memory for result! Wrong result may be returned!");
-          *out_length = 0;
-          return nullptr;
-        }
-        memcpy(result_buffer, out.data(), *out_length);
+      std::string url_string(url, query_start_idx + 1);
+      auto status = uri.Parse(url_string);
+      if (!status.ok()) {
+        return nullptr;
+      }
 
-        return result_buffer;
+      std::string query_string(url + query_start_idx + 1, url_len - query_start_idx - 1);
+      RE2 re2("(&|^)" + pattern_string + "=([^&|^#]*)");
+      int groups_num = re2.NumberOfCapturingGroups();
+      RE2::Arg *args[groups_num];
+      for (int i = 0; i < groups_num; i++) {
+        args[i] = new RE2::Arg;
+      }
+      *(args[1]) = &out;
+      // Use re2 instead of pattern_ for better performance.
+      bool matched = RE2::PartialMatchN(query_string, re2, args, groups_num);
+      if (!matched) {
+        *out_length = 0;
+        return nullptr;
       }
+
+      *out_length = static_cast<int32_t>(out.length());
+      char *result_buffer = reinterpret_cast<char *>(ctx->arena()->Allocate(*out_length));
+      if (result_buffer == NULLPTR) {
```

Comment:
Can `*out_length` be 0? If so, we should not try to allocate zero byte memory. Instead, just return a result.

Reply:
Sure